webapp/src/cloud.ts

```diff
@@ -0,0 +1,299 @@
+    // ensure we don't run this twice
+    if (_inProgressSyncAsync.isResolved()) {
+        _inProgressSyncAsync = syncAsyncInternal()
+    }
+    return _inProgressSyncAsync
+}
+
+async function syncAsyncInternal(): Promise<any> {
+    if (!auth.hasIdentity()) { return; }
+    if (!await auth.loggedIn()) { return; }
+    try {
+        pxt.log("Synchronizing with the cloud...")
+        const userId = auth.user()?.id;
+        // Filter to cloud-synced headers owned by the current user.
+        const localCloudHeaders = workspace.getHeaders(true)
+            .filter(h => h.cloudUserId && h.cloudUserId === userId);
+        const syncStart = U.nowSeconds()
+        const agoStr = (t: number) => `${syncStart - t} seconds ago`
+        const remoteHeaders = await listAsync();
+        const numDiff = remoteHeaders.length - localCloudHeaders.length
+        if (numDiff !== 0) {
+            pxt.log(`${Math.abs(numDiff)} ${numDiff > 0 ? 'more' : 'fewer'} projects found in the cloud.`);
+        }
+        const lastCloudChange = remoteHeaders.length ? Math.max(...remoteHeaders.map(h => h.modificationTime)) : syncStart
+        pxt.log(`Last cloud project change was ${agoStr(lastCloudChange)}`);
+        const remoteHeaderMap = U.toDictionary(remoteHeaders, h => h.id);
+        const localHeaderChanges: pxt.Map<Header> = {}
+        let tasks = localCloudHeaders.map(async (local) => {
+            const remote = remoteHeaderMap[local.id];
+            delete remoteHeaderMap[local.id];
+            if (remote) {
+                local.cloudLastSyncTime = remote.cloudLastSyncTime
+                // Note that we use modification time to detect differences. If we had full (or partial) history, we could
+                //  use version numbers. However we cannot currently use etags since the Cosmos list operations
+                //  don't return etags per-version. And because of how etags work, the record itself can never
+                //  have the latest etag version.
+                if (local.modificationTime !== remote.modificationTime) {
+                    const remoteFile = await getAsync(local);
+                    if (local.cloudCurrent) {
+                        // No local changes, download latest.
+                        const newHeader = {...local, ...remoteFile.header} // make sure we keep local-only metadata like _rev
+                        localHeaderChanges[remoteFile.header.id] = newHeader
+                        await workspace.saveAsync(newHeader, remoteFile.text, true);
+                    } else {
+                        // Possible conflict.
+                        const conflictStr = `conflict found for '${local.name}' (${local.id.substr(0, 5)}...). Last cloud change was ${agoStr(remoteFile.header.modificationTime)} and last local change was ${agoStr(local.modificationTime)}.`
+                        // last write wins.
+                        if (local.modificationTime > remoteFile.header.modificationTime) {
```

Comment:
Did you want to switch to _etag here (if it is being returned)?

Reply:
Honestly I'm not sure what we would do in the case where `local.modificationTime == remoteFile.header.modificationTime` and `local.cloudVersion !== remoteFile.header.cloudVersion`.
We're inherently depending on modification in the current iteration no matter what.

The List operation doesn't currently return etags so we'd need to modify the backend to support this (now that we suspect it's possible.) I'd prefer to do this in follow up PRs.